### PR TITLE
Hide row for zero categories

### PR DESCRIPTION
**Pull request by waximabbax**
For: https://github.com/ProfessionalWiki/PageApprovals/issues/101

Although I didn't make many changes, there is mainly just one function for filtering approvers, but other approaches came to my mind. 

1- It might be cleaner to have separate Endpoints for Add/Delete Categories. If we do that, we will either have to refresh the page or manipulate the HTML with JS, which is also not very clean. 

2- Another approach might be to Delete the category from DB, but when to delete it also requires building some logic. 

3- We can also keep the approver name only on the FE and never send it to the BE unless it has categories, but then we will also have to manipulate HTML using JS, which is also not the cleanest way.

Updated Test:

https://github.com/user-attachments/assets/50822668-b317-4ec6-ad1a-84207e0a1329

I request reviewers to test the UI if possible. Although I tested thoroughly but maybe I missed something. There is one case that it doesn't cover, which I mentioned here:

https://github.com/ProfessionalWiki/PageApprovals/issues/101#issuecomment-2290185591

### CODE REVIEW

File: src/EntryPoints/Specials/SpecialManageApprovers.php [hunk 1]
```diff
@@ -33,8 +33,6 @@ public function execute( $subPage ): void {
 		$request = $this->getRequest();
 		if ( $request->wasPosted() ) {
 			$this->handlePostRequest( $request );
-			$this->getOutput()->redirect( $this->getPageTitle()->getLocalURL() );
-			return;
```
Comment (JeroenDeDauw) [inline src/EntryPoints/Specials/SpecialManageApprovers.php:37]: Why this change?

File: src/EntryPoints/Specials/SpecialManageApprovers.php [hunk 2]
```diff
@@ -109,10 +107,19 @@ private function approversToViewModel( array $approvers ): array {
...
+			$user = $this->userFactory->newFromName( $request->getText( 'username' ) );
+			$showCategories = $request->wasPosted() && $user->getName() === $approver->username;
+		}
+
 		return [
 			'username' => $approver->username,
 			'userId' => $approver->userId,
-			'categories' => $approver->categories
+			'categories' => $approver->categories,
+			'showCategories' => $showCategories
```
Comment (JeroenDeDauw) [inline src/EntryPoints/Specials/SpecialManageApprovers.php:122]: This is poorly named. It's not indicating if categories should be shown

File: src/EntryPoints/Specials/SpecialManageApprovers.php [hunk 3]
```diff
@@ -109,10 +107,19 @@ private function approversToViewModel( array $approvers ): array {
 	 * @return array<string, mixed>
 	 */
 	private function approverToViewModel( Approver $approver ): array {
+		$showCategories = true;
+
+		if ( empty( $approver->categories ) ) {
+			$request = $this->getRequest();
+			$user = $this->userFactory->newFromName( $request->getText( 'username' ) );
+			$showCategories = $request->wasPosted() && $user->getName() === $approver->username;
+		}
```
Comment (JeroenDeDauw) [inline src/EntryPoints/Specials/SpecialManageApprovers.php:116]: This can be in its own function returning a bool

And you can do it without mutation (assigning to an already assigned variable)

File: src/EntryPoints/Specials/SpecialManageApprovers.php (hunk 2 above)
Comment (JeroenDeDauw) [reply src/EntryPoints/Specials/SpecialManageApprovers.php:122]: Though why output this data structure at all, to then not use it? I think it makes more sense to filter out the Approver without categories from the Approver[]. No changes needed on the frontend. Also easier to test

File: src/EntryPoints/Specials/SpecialManageApprovers.php (hunk 3 above)
Comment (JeroenDeDauw) [reply src/EntryPoints/Specials/SpecialManageApprovers.php:116]: Took me a bit to understand why we have this logic at all. It's because using the UI causes post requests which reload the page, so without this logic, the row would vanish as soon as you delete the last category, correct? 

(Having a well names private function would make this more obvious)

Comment (waximabbax) [reply src/EntryPoints/Specials/SpecialManageApprovers.php:116]: Yes. That's right.

File: src/EntryPoints/Specials/SpecialManageApprovers.php (hunk 1 above)
Comment (waximabbax) [reply src/EntryPoints/Specials/SpecialManageApprovers.php:37]: This is used to redirect the user back to the special page. We needed this because if we didn't, the user would keep sending the POST request upon every refresh or page change.

However; if we do it right now, we will lose context, and we won't know if the Action was 'Add-Approver' or 'Delete' or simply we won't know if it was a POST request since for every POST request we force a GET request using 

`$th.is->getOutput()->redirect( $this->getPageTitle()->getLocalURL() );`

A small JS script is taking care of resending POST requests upon page refresh.

**Review by malberts — CHANGES_REQUESTED** [review]
I am still busy testing, but ran into a somewhat bad issue:
* one user with categories
* add another user, but don't add categories
* add another user
* everyone disappears (first user comes back after page reload)

[disappear.webm](https://github.com/user-attachments/assets/7b623431-f5c0-42bf-94ad-010841033457)

**Comment by malberts** [comment]
Blocker:
* have two users with categories
* remove the category for the first user
* the page reloads with the first user still having no category
* reload page
* everything disappears
* adding new user no longer works

[disappear2.webm](https://github.com/user-attachments/assets/6fa77d57-77ad-4c80-bf5f-136ff1f42d4f)

File: src/EntryPoints/Specials/SpecialManageApprovers.php [hunk 4]
```diff
@@ -69,15 +67,13 @@ private function processCategoryAction( string $action, string $category, int $u
 		$currentCategories = $this->approverRepository->getApproverCategories( $userId );
 
 		switch ( $action ) {
+			case 'add-approver':
```
Comment (waximabbax) [inline src/EntryPoints/Specials/SpecialManageApprovers.php:70]: 1- We can remove the `add-approver` case. We can add the approver if it doesn't exist using `add` case, 

2-and also, the reason for moving it here now is that if the user accidentally adds the user that's already in the DB with many categories, it deletes all of the categories.

The reason for keeping `add-approver'` on the FE is to show the intent of the action/request. I can remove it though.

File: src/EntryPoints/Specials/SpecialManageApprovers.php [hunk 5]
```diff
@@ -99,10 +95,33 @@ private function renderHtml( array $approversCategories ): void {
...
+	}
+
+	/**
+	 * @param array<Approver> $approvers
+	 *
 	 * @return array<array<string, mixed>>
 	 */
 	private function approversToViewModel( array $approvers ): array {
-		return array_map( fn( Approver $approver ) => $this->approverToViewModel( $approver ), $approvers );
+		return array_values(
```
Comment (JeroenDeDauw) [inline src/EntryPoints/Specials/SpecialManageApprovers.php:119]: Why the added array_values? For the tests?

File: tests/EntryPoints/Specials/SpecialManageApproversTest.php [hunk 6]
```diff
@@ -64,18 +67,13 @@ public function testAddApproverAction(): void {
...
-		$this->assertStringNotContainsString(
-			'<div class="category-entry">',
-			$output,
-			'New approver should have no categories'
-		);
+		$approver = array_filter( $approvers, fn( $approver ) => $approver['userId'] === $userId );
+		$newApprover = reset( $approver );
+
+		$this->assertNotNull( $newApprover, 'New approver should be added' );
+		$this->assertSame( [], $newApprover['categories'], 'New approver should have no categories' );
```
Comment (JeroenDeDauw) [inline tests/EntryPoints/Specials/SpecialManageApproversTest.php:76]: This drew my attention since logic such as the array_filter is a yellow flag in tests.

It seems OK in this case, since we do not have a proper ApproverList collection class that could house this logic. However it's not nice to pollute the test case with these lower level details. You need to read all of them to guess at the intent. Better to extract a well named function. Perhaps `assertNewApproverExists( int $userId ): void`, `getApproverWithUserId( int $userId ): ?Approver`, or some combination.

File: src/EntryPoints/Specials/SpecialManageApprovers.php (hunk 5 above)
Comment (waximabbax) [reply src/EntryPoints/Specials/SpecialManageApprovers.php:119]: This was causing an issue that @malberts reported above where deleting an older user made some rows disappear.

1- Our approversToViewModel method was returning an associative array where the keys were preserved from the original data structure. This resulted in an array like this:

```
array(1) {
  [1] => array(3) {
    ["username"] => string(10) "AdminName3"
    ["userId"] => int(4)
    ["categories"] => array(1) { ... }
  }
}
```

2- Problem with Mustache:
Mustache's {{#approvers}} loop expects an array with numeric indexes starting from 0. It doesn't work well with associative arrays or arrays with non-sequential numeric keys.

3- Solution:

- array_values() returns all the values from the array and indexes them numerically.
- This new structure is exactly what Mustache expects for its loops.
- It ensures that no matter how the original data is indexed, we always pass a cleanly indexed array to the template.

By adding array_values(), we convert the associative array into an indexed array which would print something like this:

```
array(1) {
  [0] => array(3) {
    ["username"] => string(10) "AdminName3"
    ["userId"] => int(4)
    ["categories"] => array(1) { ... }
  }
}
```

**Review by JeroenDeDauw — APPROVED** [review]
I did a manual test as well, but would be nice if @malberts can repeat.

File: src/EntryPoints/Specials/SpecialManageApprovers.php (hunk 5 above)
Comment (JeroenDeDauw) [reply src/EntryPoints/Specials/SpecialManageApprovers.php:119]: Right, so we needed an integration test that looks at the output of mustache.

Turns out that building the HTML on the backend created several challenges. For instance, we now have a request that adds an approver with no categories that is then persisted, which on its own makes little sense, but is needed for the UI to show correctly. Perhaps building the UI with JS and sending background POSTs would have been better.

**Comment by malberts** [comment]
Both my isses are fixed.

I am also able to reproduce the behavior where adding multiple users without categories causes the new user to replace the last empty user:
[replace-last.webm](https://github.com/user-attachments/assets/c9f39f68-d782-4300-84bc-45edb38a0f09)

However, this was mentioned in https://github.com/ProfessionalWiki/PageApprovals/issues/101#issuecomment-2290185591, so this is not an problem in this PR.
